### PR TITLE
Check routes before generating component url

### PR DESCRIPTION
**Pull request by core23**
If you disable the show or edit route in an admin, the timeline would throw an exception.

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Could you please squash your commit?

Which exception is throw? Could you elaborate the original issue by a concrete case with code sample?

Any issue related?

Thanks.

**Comment by core23** [comment]
After removing the edit route for an admin, the DefaultRouteGenerator would throw a "RuntimeException: unable to find the route `acme.demo.admin.demo.edit`" exception.

**Comment by Soullivaneuh** [comment]
I don't really like this way.

With your method, if the route does not exists, another unwanted route will be generated with no error.

This will be confused for developer. A thrown error would be more explicit.

**Comment by core23** [comment]
@Soullivaneuh 
An error would not be the best idea in the timeline. If you create some entity which has no edit route, the timeline would crash. So the route must be checked before generating edit. This fallback routing could also be done in the AdminBundle.

I might remove the list route and replace it with a simple text without a link, but the proximate used route after edit would be the show route.

**Comment by Soullivaneuh** [comment]
ping @rande for opinion.

**Comment by rande** [comment]
@Soullivaneuh the route might become available or removed depends on project life cycle. So I guess it is better to check route at runtime.

**Comment by Soullivaneuh** [comment]
I understand. But should we fallback to another route or just render a text item?

**Comment by rande** [comment]
@Soullivaneuh no idea, I think fallback to the show action is a good option and then text item

File: Twig/Extension/SonataTimelineExtension.php [hunk 1]
```diff
@@ -65,8 +65,19 @@ public function generateLink(ComponentInterface $component, ActionInterface $act
             return $component->getHash();
         }
 
-        return sprintf('<a href="%s">%s</a>',
-            $admin->generateObjectUrl('edit', $component->getData()),
+        foreach (array('edit', 'show') as $mode) {
+            if ($admin->hasRoute($mode)) {
```
Comment (rande) [inline Twig/Extension/SonataTimelineExtension.php]: You also need to check if the user can access the route

**Comment by core23** [comment]
Removed the list route and changed it to plain text

**Comment by core23** [comment]
ping @Soullivaneuh

**Comment by OskarStark** [comment]
ping @Soullivaneuh @rande looks linke this PR is ready

**Comment by Soullivaneuh** [comment]
Thank you @core23.

**Pull request merged**